core/src/main/java/org/apache/calcite/rex/RexBuilder.java

```diff
@@ -1355,10 +1356,14 @@ public RexNode makeIn(RexNode arg, List<? extends RexNode> ranges) {
     if (areAssignable(arg, ranges)) {
       final Sarg sarg = toSarg(Comparable.class, ranges, RexUnknownAs.UNKNOWN);
       if (sarg != null) {
-        final RexNode range0 = ranges.get(0);
+        final List<RelDataType> types = ranges.stream()
+            .map(RexNode::getType)
+            .collect(Collectors.toList());
+        RelDataType searchType = typeFactory.leastRestrictive(types);
+        searchType = searchType == null ? ranges.get(0).getType() : searchType;
```

Comment:
If`searchType == null` is `true`, do we really need to proceed with `ranges.get(0).getType()`?

Reply:
All the usages of `RelDataTypeFactory.leastRestrictive` handles the edge case where the return type is null. So I thought about defaulting to how it get handles now in current master.  Is it really not necessary to handle this case? Can you provide your thoughts on this?

Reply:
The question is: "is it correct to use `ranges.get(0).getType()`?". If it's not correct, then we'd better throw exception. If we decide to keep it, then we should know in what cases will `searchType == null` is true, and we still want to use `ranges.get(0).getType()` as the final type.

Reply:
Thanks @libenchao .  I think if we don't find a  least restrictive `searchType`  among the `types`, then selecting `ranges.get(0).getType()` is wrong.  Since it might eventually fail at Rel to Sql conversion.  Hence throwing an exception